sycl/include/CL/sycl/accessor.hpp

```diff
@@ -2015,8 +2015,10 @@ accessor(buffer<DataT, Dimensions, AllocatorT>, handler, Type1, Type2, Type3,
 /// \ingroup sycl_api_acc
 template <typename DataT, int Dimensions, access::mode AccessMode,
           access::placeholder IsPlaceholder>
-class __SYCL_SPECIAL_CLASS accessor<DataT, Dimensions, AccessMode,
-                                    access::target::local, IsPlaceholder> :
+class __SYCL_SPECIAL_CLASS __SYCL2020_DEPRECATED(
+    "'accessor' with 'target::local' is deprecated: use 'local_accessor' "
+    "instead.") accessor<DataT, Dimensions, AccessMode, access::target::local,
+                         IsPlaceholder> :
```

Comment:
Unfortunately [[deprecated]] attribute doesn't work on template specializations: https://godbolt.org/z/nx56bsfjz
Deprecation message is not generated.

Another problem is that even if you deprecate this template specialization then I believe usage of local_accessor alias will result in deprecated message as well, isn't it?

Reply:
Okay, that is a bit of an issue. I think a possible solution could be to deprecate the class constructors instead of the class itself. 
The only way I can think of to resolve the alias of `local_accessor` producing a deprecated message is to create a new class just for `local_accessor`.

Reply:
Indeed this seems to be very unpleasant issue, I'm sorry.

Other possible option is just to not add [[deprecated]] attribute in this patch if we are ok with absence of deprecation message. @steffenlarsen Do you think it is ok to not emit deprecation warning in this case? Since emitting a deprecation warning in this case causes too much headache and it seems to be only usability issue.

Reply:
I would definitely prefer that we deprecate the API explicitly, but you are right that we cannot deprecate specializations.

That said, SYCL 2020 deprecated select members of `target`, `local` included, so what I suggest is:

1. Deprecate `access::target::local`.
2. Move the implementation in this specialization into a new `local_accessor` class.
3. Make `local_accessor` a base class of this `accessor` specialization.

This has the added benefit of making it easy to remove the deprecated specialization when we remove `(access::)target::local` in the future. 😄

Reply:
Sounds great to me, thanks a lot for the help!

Reply:
I have updated this PR to follow the suggested approach. It takes a slightly different structure by using a `local_accessor_base` class to implement both `local_accessor` and `accessor`. This is necessary to handle `accessors` template arguments.

Additionally, some changes to Clang's SemaSYCL were applied so the new `local_accessor` is identified and correctly handled.